src/main/scala/mesosphere/marathon/core/storage/PersistenceStore.scala

```diff
@@ -0,0 +1,254 @@
+    */
+  def get[Id, V](id: Id)(implicit
+    ir: IdResolver[Id, K, V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = {
+
+    async {
+      await(rawGet(ir.toStorageId(id))) match {
+        case Some(v) =>
+          Some(await(um(v)))
+        case None =>
+          None
+      }
+    }
+  }
+
+  /**
+    * Create the value for the given Id, failing if it already exists
+    *
+    * @return A Future that will complete once the value has been created at the given id, or fail with
+    *         [[StoreCommandFailedException]] if either a value already exists at the given Id, or
+    *         if there is an underlying storage problem
+    */
+  def create[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, K, V, Serialized],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Done] = {
+    createOrUpdate(id, (oldValue: Option[V]) => oldValue match {
+      case None => Success(v)
+      case Some(existing) =>
+        Failure(new StoreCommandFailedException(s"Unable to create $id as it already exists ($existing)"))
+    }).map(_ => Done)(CallerThreadExecutionContext.callerThreadExecutionContext)
+  }
+
+  /**
+    * Update the value at the given Id if and only if the 'change' method returns a [[scala.util.Try]], if
+    * the try is successful, the value will be updated.
+    * A value must already exist at the given id.
+    *
+    * @return A Future that will complete with the previous value if there was already an
+    *         existing value, even if the change
+    *         callback fails.
+    *         Any other condition should fail the future with [[StoreCommandFailedException]].
+    */
+  def update[Id, V](id: Id)(change: V => Try[V])(implicit
+    ir: IdResolver[Id, K, V, Serialized],
+    um: Unmarshaller[Serialized, V],
+    m: Marshaller[V, Serialized]): Future[V] = {
+    createOrUpdate(id, (oldValue: Option[V]) => oldValue match {
+      case Some(old) =>
+        change(old)
+      case None =>
+        Failure(new StoreCommandFailedException(s"Unable to update $id as it doesn't exist"))
+    }).map(_.get)(CallerThreadExecutionContext.callerThreadExecutionContext)
+  }
+
+  /**
+    * Delete the value at the given Id, idempotent
+    *
+    * @return A future indicating whether the value was deleted (or simply didn't exist). Underlying storage issues
+    *         will fail the future with [[StoreCommandFailedException]]
+    */
+  def delete[Id, V](k: Id)(implicit ir: IdResolver[Id, K, V, Serialized]): Future[Done] = {
+    rawDelete(ir.toStorageId(k))
+  }
+
+  /**
+    * @return A source of _all_ keys in the Persistence Store (which can be used by a
+    *         [[mesosphere.marathon.core.storage.impl.CachingPersistenceStore]] to populate the
+    *         cache completely on startup.
+    */
+  protected[storage] def keys(): Source[K, NotUsed]
```

Comment:
Perhaps divide this trait into:
- an abstract with public methods trait without any implementation
- an abstract class
This will also help with implementations (like cache) that does not need the raw* methods.